cylc/flow/network/__init__.py

```diff
@@ -14,3 +14,163 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """Package for network interfaces to cylc suite server objects."""
+
+import asyncio
+from threading import Thread
+from time import sleep
+
+import zmq
+import zmq.asyncio
+
+from cylc.flow import LOG
+from cylc.flow.exceptions import CylcError
+
+
+class ZMQSocketBase:
+    """Initiate the ZMQ socket bind for specified pattern on new thread.
+
+    NOTE: Security to be provided via zmq.auth (see PR #3359).
+
+    Args:
+        pattern (enum): ZeroMQ message pattern (zmq.PATTERN).
+
+        context (object, optional): instantiated ZeroMQ context, defaults
+            to zmq.asyncio.Context().
+
+        barrier (object, optional): threading.Barrier object for syncing with
+            other threads.
+
+        threaded (bool, optional): Start socket on separate thread.
+
+        daemon (bool, optional): daemonise socket thread.
+
+    This class is designed to be inherited by REP Server (REQ/REP)
+    and by PUB Publisher (PUB/SUB), as the start-up logic is similair.
+
+
+    To tailor this class overwrite it's method on inheritance.
+
+    """
+
+    def __init__(self, pattern, bind=False, context=None,
+                 barrier=None, threaded=False, daemon=False):
+        self.bind = bind
+        if context is None:
+            self.context = zmq.asyncio.Context()
```

Comment:
Should we create a new Context here, or call `Context.instance()` to return the global context? Not sure if that's because of the threads... but what happens if somewhere within the same thread we call this constructor twice with no context? As far as I know, it's safe to have multiple contexts, though bad for performance? (only necessary for IPC communication I think, which we don't use I guess)

Reply:
We instantiate a context at application level, and pass it into the constructor.. It is advised by ZeroMQ to only have one context per application, unless there's reason not to (I think it means we can close the context and associated sockets (?)). A ZeroMQ context is thread safe. (sockets are not). We only construct a context there if it's not passed in, I'm not calling it twice (if I did there wouldn't be a problem).